conn_browser.go

```diff
@@ -0,0 +1,311 @@
+// +build js,wasm
+
+package websocket
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"syscall/js"
+	"time"
+)
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errConnectionClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler  *js.Func
+	closeHandler    *js.Func
+	errorHandler    *js.Func
+	mut             sync.Mutex
+	currDataMut     sync.RWMutex
+	currData        bytes.Buffer
+	closeOnce       sync.Once
+	closeSignalOnce sync.Once
+	closeSignal     chan struct{}
+	dataSignal      chan struct{}
+	localAddr       net.Addr
+	remoteAddr      net.Addr
+	firstErr        error
+}
+
+// NewConn creates a Conn given a regular js/wasm WebSocket Conn.
+func NewConn(raw js.Value) *Conn {
+	conn := &Conn{
+		Value:       raw,
+		closeSignal: make(chan struct{}),
+		dataSignal:  make(chan struct{}, 1),
+		localAddr:   NewAddr("0.0.0.0:0"),
+		remoteAddr:  getRemoteAddr(raw),
+	}
+	// Force the JavaScript WebSockets API to use the ArrayBuffer type for
+	// incoming messages instead of the Blob type. This is better for us because
+	// ArrayBuffer can be converted to []byte synchronously but Blob cannot.
+	conn.Set("binaryType", "arraybuffer")
+	conn.setUpHandlers()
+	return conn
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	if err := c.checkOpen(); err != nil {
+		return c.readAfterErr(b)
+	}
+
+	for {
+		c.currDataMut.RLock()
+		n, _ := c.currData.Read(b)
+		c.currDataMut.RUnlock()
+
+		if n != 0 {
+			// Data was ready. Return the number of bytes read.
+			return n, nil
+		}
+
+		// There is no data ready to be read. Wait for more data or for the
+		// connection to be closed.
+		select {
+		case <-c.dataSignal:
+			continue
+		case <-c.closeSignal:
+			return c.readAfterErr(b)
+		}
+	}
+}
+
+// readAfterError reads from c.currData. If there is no more data left it
+// returns c.firstErr if non-nil and otherwise returns io.EOF.
+func (c *Conn) readAfterErr(b []byte) (int, error) {
+	c.currDataMut.RLock()
+	n, err := c.currData.Read(b)
+	c.currDataMut.RUnlock()
+	if n == 0 {
+		if c.firstErr != nil {
+			return 0, c.firstErr
+		} else {
+			return 0, io.EOF
+		}
+	}
+	return n, err
+}
+
+// checkOpen returns an error if the connection is not open. Otherwise, it
+// returns nil.
+func (c *Conn) checkOpen() error {
```

Comment:
Can we just wait for the "close" event to happen then check the `closeSignal` channel after that? That should save us a round-trip through the webasm runtime.

```go
select {
  case <-c.closeSignal:
    return true
  default:
    return false
}
```